libc/dirent/readdir.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * test/test_strftime.c
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Arkadiusz Kozlowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <time.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <limits.h>
+#include <signal.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <unity_fixture.h>
+
+#include "common.h"
+
+#define MAIN_DIR            "test_readdir"
+#define INO_T_TEST_MAX_DIRS 10
+
+int d_ino_in(ino_t arg, ino_t *arr)
+{
+	for (int i = 0; i < INO_T_TEST_MAX_DIRS; ++i)
+		if (arg == arr[i])
+			return i + 1;
+	return 0;
+}
+
+
+TEST_GROUP(dirent_readdir);
+
+TEST_SETUP(dirent_readdir)
+{
+	mkdir(MAIN_DIR, 0777);
+
+	mkdir(MAIN_DIR "/dir1", 0777);
+	mkdir(MAIN_DIR "/dir2", 0777);
+
+	mkdir(MAIN_DIR "/dir1/nest1", 0777);
+	mkdir(MAIN_DIR "/dir1/nest2", 0777);
+
+	mkdir(MAIN_DIR "/dir2/nest1", 0777);
+	mkdir(MAIN_DIR "/dir2/nest2", 0777);
+
+	FILE *files[] = {
+		fopen(MAIN_DIR "/file1.txt", "w+"),
+		fopen(MAIN_DIR "/file2.dat", "w+"),
+		fopen(MAIN_DIR "/file3.json", "w+")
+	};
+
+	if (files[0]) {
+		fprintf(files[0], "Some data");
+		fprintf(files[1], "Some other data");
+
+
+		fclose(files[0]);
+		fclose(files[1]);
+		fclose(files[2]);
+	}
+}
+
+
+TEST_TEAR_DOWN(dirent_readdir)
+{
+	rmdir(MAIN_DIR "/dir1/nest1");
+	rmdir(MAIN_DIR "/dir1/nest2");
+
+	rmdir(MAIN_DIR "/dir2/nest1");
+	rmdir(MAIN_DIR "/dir2/nest2");
+
+	rmdir(MAIN_DIR "/dir1");
+	rmdir(MAIN_DIR "/dir2");
+
+	remove(MAIN_DIR "/file1.txt");
+	remove(MAIN_DIR "/file2.dat");
+	remove(MAIN_DIR "/file3.json");
+
+	rmdir(MAIN_DIR);
+}
+
+
+TEST(dirent_readdir, long_name_directory_check)
+{
+	errno = 0;
+	DIR *dp = opendir(MAIN_DIR);
+	struct dirent *info;
+
+	char longDirName[NAME_MAX + 1] = { 0 };
+	longDirName[NAME_MAX] = '\0';
+
+	memset(longDirName, 'a', NAME_MAX);
+	char longDirPath[NAME_MAX + 2 + sizeof(MAIN_DIR)];
+	sprintf(longDirPath, MAIN_DIR "/%s", longDirName);
+
+	if (mkdir(longDirPath, 0777) == -1)
+		TEST_IGNORE_MESSAGE(strerror(errno));
+
+	int passed = 0;
+
+	while (info = readdir(dp)) {
+		if (!strcmp(longDirName, info->d_name)) {
+			closedir(dp);
+			rmdir(longDirPath);
+			TEST_PASS();
+		}
+	}
+
+	closedir(dp);
+	rmdir(longDirPath);
+	TEST_FAIL();
+}
+
+
+TEST(dirent_readdir, basic_listing_count)
+{
+	DIR *dp = opendir(MAIN_DIR);
+	int entry_counter = 0;
+	struct dirent *info;
+
+	if (!dp)
+		TEST_FAIL_MESSAGE(strerror(errno));
+
+	while (info = readdir(dp))
+		entry_counter++;
+
+	/* 6 files from setup, and both . and .. directories */
+	TEST_ASSERT_EQUAL(7, entry_counter);
+
+	closedir(dp);
+}
+
+
+TEST(dirent_readdir, reading_in_parent_and_child)
```

Comment:
This tastcase missed completly its puprose. You should create 1 directory stream then `fork()` and use `readdir()` in one of them, preferably test reading once in child and then recreate setup and read in parent.

Reply:
That's almost exactly what I did, the only difference is that I created two pointers, one for parent and second one for child, and checked each of them separately in each of the processes.